File: server/app/controller/auth_controller.py
from fastapi import APIRouter
from fastapi import Response, HTTPException, Request
from app.schema.auth_schema import LoginUser, RegisterUser
from app.service.auth_service import AuthService


auth_router = APIRouter()


@auth_router.post("/login")
async def login(user: LoginUser, response: Response) -> dict[str, str]:
    result = AuthService.login(user.email, user.password)
    if result:
        response.set_cookie(
            key="access_token", value=result["access_token"], httponly=True
        )
        response.set_cookie(
            key="refresh_token", value=result["refresh_token"], httponly=True
        )
        return {"message": "Login is successful."}
    else:
        raise HTTPException(status_code=500, detail="Server error!")


@auth_router.post("/register", status_code=201)
async def register(user: RegisterUser):
    user = AuthService.register(payload=user)
    if user:
        return {"message": "User has been created."}


@auth_router.post("/logout")
async def logout(response: Response) -> dict[str, str]:
    response.delete_cookie("access_token")
    response.delete_cookie("refresh_token")
    return {"message": "you are logged out."}

File: node-server/src/controllers/auth-controller.ts
import { Response, Request } from 'express';
import { AuthService } from '../services/auth-service';
import { Exception } from '../exceptions/exception';

export class AuthController {
    private authService: AuthService;

    constructor() {
        this.authService = new AuthService();
    }

    public login = async (req: Request, res: Response) => {
        try {
            const { email, password } = req.body;
            const result = await this.authService.login(email, password);
            res.cookie('access_token', result.access_token, {
                httpOnly: true,
            });
            res.cookie('refresh_token', result.refresh_token, {
                httpOnly: true,
            });
            res.status(200).json({ message: 'Login is successful.' });
        } catch (error) {
            if (error instanceof Exception) {
                res.status(error.statusCode).json({ message: error.message });
            } else {
                res.status(500).json({ message: 'Internal server error!' });
            }
        }
    };

    public register = async (req: Request, res: Response) => {
        try {
            await this.authService.register(req.body);
            res.status(201).json({ message: 'Account created successfully.' });
        } catch (error) {
            if (error instanceof Exception) {
                res.status(error.statusCode).json({ message: error.message });
            } else {
                res.status(500).json({ message: 'Internal server error!' });
            }
        }
    };

    public verify = async (req: Request, res: Response) => {
        try {
            const token: string = req.cookies.access_token;
            res.status(200).json({ status: true, user: await this.authService.verify(token) });
        } catch (error) {
            if (error instanceof Exception) {
                res.status(error.statusCode).json({ message: error.message });
            } else {
                res.status(500).json({ message: 'Internal server error!' });
            }
        }
    };

    public logout = async (req: Request, res: Response) => {
        res.clearCookie('access_token');
        res.clearCookie('refresh_token');
        res.status(200).json({ message: 'User logout successful.' });
    };

    public updatePassword = async (req: Request, res: Response) => {
        const { currentPassword, newPassword } = req.body;

        try {
            const token: string = req.cookies.access_token;
            const result = await this.authService.updatePassword(
                token,
                currentPassword,
                newPassword,
            );
            res.status(200).json({ message: result.message });
        } catch (error) {
            if (error instanceof Exception) {
                res.status(error.statusCode).json({ message: error.message });
            } else {
                res.status(500).json({ message: 'Internal server error!' });
            }
        }
    };
}
